### DISCUSSION TITLE
RasterMath

### DISCUSSION THREAD

**Discussion by grantboxer** [EnMAP-Box/enmap-box#590] [General]
I am trying to use code snippets in the raster math dialog box but the link to the cookbook is broken "https://enmap-box.readthedocs.io/en/latest/usr_section/usr_cookbook/raster_math.html".
I would like to get more information on writing the code snippets.
Cheers Grant

**Comment by janzandr** [comment]
Thanks for reporting, will be fixed. Correct link is:
https://enmap-box.readthedocs.io/en/latest/usr_section/usr_manual/applications.html#raster-math

**Comment by grantboxer** [comment]
Hi Andreas,

 

When I try to run a raster math via the code snippets I get the following message,

 

Any thoughts?

 

Cheers Grant

 

2023-09-05T08:56:13     WARNING    warning::4: RuntimeWarning: invalid value encountered in divide
             
             traceback: File "C:\Users\Grant\AppData\Roaming\Python\Python39\site-packages\typeguard\__init__.py", line 1033, in wrapper
              retval = func(*args, **kwargs)
              File "C:\Users/Grant/AppData/Roaming/QGIS/QGIS3\profiles\Grant1/python/plugins\enmapboxplugin\enmapboxprocessing\algorithm\rastermathalgorithm\rastermathalgorithm.py", line 295, in processAlgorithm
              results = self.processBlock(
              File "C:\Users\Grant\AppData\Roaming\Python\Python39\site-packages\typeguard\__init__.py", line 1033, in wrapper
              retval = func(*args, **kwargs)
              File "C:\Users/Grant/AppData/Roaming/QGIS/QGIS3\profiles\Grant1/python/plugins\enmapboxplugin\enmapboxprocessing\algorithm\rastermathalgorithm\rastermathalgorithm.py", line 567, in processBlock
              exec(code, namespace)
              File "", line 4, in 
              File "C:\OSGeo4W\apps\Python39\lib\warnings.py", line 109, in _showwarnmsg
              sw(msg.message, msg.category, msg.filename, msg.lineno,

**Comment by janzandr** [comment]
Probably a divide by zero in one of the pixels. Usually nothing to worry, if the result raster is looking fine. Zero divides often occure for no data pixel.

**Comment by janzandr** [comment]
What was the code snippet?

**Comment by grantboxer** [comment]
Thanks Andreas,

 

I cannot seem to get it to run a code snippet.

 

{'raster': QgsRasterLayer}

 

b206 = ***@***.***

b207 = ***@***.***

p2300 = b206/b207

outputRaster = p2300

 

Any thoughts on the required syntax?

 

Cheers Grant

**Comment by janzandr** [comment]
Oh, I thought it is running fine and you are just wondering about the warning (which isn't critical).
Why isn't it running?
Do you get another error message?
Isn't the result raster not visible inside the DataSources panel?
Please show a screenshot.

**Comment by grantboxer** [comment]
Hi Andreas,

 

I have used the ndvi code snippet as a proforma but I still get this message,

 

2023-09-05T16:09:07     WARNING    warning::4: RuntimeWarning: invalid value encountered in divide
             
             traceback: File "C:\Users\boxer\AppData\Roaming\Python\Python39\site-packages\typeguard\__init__.py", line 1033, in wrapper
              retval = func(*args, **kwargs)
              File "C:\Users/boxer/AppData/Roaming/QGIS/QGIS3\profiles\default/python/plugins\enmapboxplugin\enmapboxprocessing\algorithm\rastermathalgorithm\rastermathalgorithm.py", line 295, in processAlgorithm
              results = self.processBlock(
              File "C:\Users\boxer\AppData\Roaming\Python\Python39\site-packages\typeguard\__init__.py", line 1033, in wrapper
              retval = func(*args, **kwargs)
              File "C:\Users/boxer/AppData/Roaming/QGIS/QGIS3\profiles\default/python/plugins\enmapboxplugin\enmapboxprocessing\algorithm\rastermathalgorithm\rastermathalgorithm.py", line 567, in processBlock
              exec(code, namespace)
              File "", line 4, in 
              File "C:\OSGeo4W\apps\Python39\lib\warnings.py", line 109, in _showwarnmsg
              sw(msg.message, msg.category, msg.filename, msg.lineno,
             

When I run this,

 

'raster': QgsRasterLayer}

 

# find bands

A = ***@***.***

B = ***@***.***

 

# calculate p2300

p2300 = A / B

 

# mask no data region

noDataValue = -9999

***@***.*** = noDataValue

***@***.*** = noDataValue

 

# set no data value and band name

p2300.setNoDataValue(noDataValue)

p2300.setBandName('NDVI', bandNo=1)

 

# clean up temp variables

del A, B

 

Where has I gone wrong? I am using a PRISMA file.

 

Cheers Grant

**Comment by janzandr** [comment]
> 2023-09-05T16:09:07     WARNING    warning::4: RuntimeWarning: invalid value encountered in divide

This is just a warning, which you can ignore for now.
Any other error messagen?

1. Please you a screenshot of the **Raster math** dialog before you click the **Run** button.
2. Also show a screenshot after the algorithm has finished.

**Comment by grantboxer** [comment]
Screenshot before running

 



 

 

Screenshot after running

 



 

Hope that helps!

 

Cheers Grant

**Comment by janzandr** [comment]
Hi @grantboxer, the screenshots aren't displayed.
In case you are replying via email directly, I would ask you to go on GitHub and repost.
Generally, it would be better if you wouldn't use the reply via email option, because it introduces a lot of clutter to the discussion.
Thanks and regards
Andreas

**Reply by grantboxer** [reply]
OK, will try again, hopefully this works.
Before running
![image](https://github.com/EnMAP-Box/enmap-box/assets/21163329/d0421a77-e91e-41ab-b548-098ab1df1c9c)

After running
![image](https://github.com/EnMAP-Box/enmap-box/assets/21163329/e53bff49-2f93-40f6-bd53-5d8ffa75981d)

**Comment by janzandr** [comment]
Alright, tanks.
Your output is named **p2300**, so you have to choose an output filename that uses the same basename, e.g.
![image](https://github.com/EnMAP-Box/enmap-box/assets/45030715/2b6f2a33-44c0-49b5-8c2f-3ad6416b70e0)

And yes, I agree, the error message is confusing:
![image](https://github.com/EnMAP-Box/enmap-box/assets/45030715/c8c50b0c-48df-42cb-a1cb-7c68938d60d4)

Will fix this.

**Reply by grantboxer** [reply]
Thanks, that works but it is a little clunky if we have to be specific for each output raster.
Can we do a series of calculations to create a number of output raster bands, or can we only do one at a time?
I would like to calculate about 10 expressions in the one script if possible.

**Comment by janzandr** [comment]
Sure, you can calculate 10 outputs in one script. Additional outputs will be placed in the same folder as the default output (i.e. the you specify the name for)

And yes, agreed, it is a bit clunky. I'll discuss this with @jakimowb and team again, if we can come up with a better design for specifying output filenames for multiple outputs.

**Comment by jakimowb** [comment]
The algorithm could show that it is well-parameterized before execution.
For example, [`checkParameterValues(...)`](
https://api.qgis.org/api/classQgsProcessingAlgorithm.html#ae3991f84be1b939144df415bec29d3fd) could check if the choosen output names appear in the provied code snippets.
The algorithm could already fail before it's execution to show that it

**Comment by janzandr** [comment]
It's a bit more complicated. The algo isn't failing, all results are written, but if the default output isn't used (which is fine) an error is raised.
A simple solution would be to make the default output optional.

**Reply by jakimowb** [reply]
In this case, who raises the error if the output is not required at all?

**Comment by janzandr** [comment]
The bug is, that the default output parameter is currently not flagged as optional.
Simple solution: flag the parameter as optional :-)

**Comment by janzandr** [comment]
Problem solved: https://github.com/EnMAP-Box/enmap-box/issues/615
